### ISSUE TITLE
Implement XMLRPC call to Anope on user registration/password change

### ISSUE THREAD

**Issue by kenneaal** [FuelRats/api.fuelrats.com#84]
The API can call Anope's XMLRPC endpoint to register a nickname on the IRCD for the user when they are created, or change its password when they change it in the API.

This requires the addition of a Nickname field to the registration form, which can be stored in the user model (where it already has a field).
The XMLRPC call is a query to http://ServicesHost:port/xmlrpc. The encoding of the request should be text/xml, and can be a text encoded string faking the XML, or an XMLRPC constructed request.
The XML should look like this:

`
<?xml version="1.0" encoding="iso-8859-1"?>
<methodCall>
<methodName>command</methodName>
<params>
 <param>
  <value>
   <string>Nickserv</string>
   <string>**NICKNAME**</string>
   <string>register **PASSWORD** **EMAILADDRESS**</string>
  </value>
 </param>
</params>
</methodCall>
`
Basically, the first string is the service accessed (Nickserv), second is the executor (Who should we emulate the call comes from) and third is the arguments.

**Comment by trezy** [comment]
Awesome! This should be pretty easy, too: https://www.npmjs.com/package/xml2json

**Comment by xlexi** [comment]
<Absolver> Well, it's gonna be XMLRPC, then. And the way I envision it is that when the profile page is displayed, the value in the nickname (Could be called primary nickname) is checked against the nickserv database through XMLRPC.
<Absolver> If the nick doesn't exist, present a "Register nick" button that allows it to be registered in nickserv, with same password as account.
<Absolver> If the nick DOES exist, there needs to be a way to ensure that they're not trying to steal someone else's nick, since the API would have blanket permission to reset the password of ANY nick on the server.

**Comment by xlexi** [comment]
Assigned to DerryBear, expect results in 3 days

**Comment by xlexi** [comment]
@kenneaal  if the website and IRC passwords are to be synchronised the API needs a way to know when the user changes it via IRC.
How will we go about this?

**Comment by kenneaal** [comment]
@xlexi Well, we can't get nickserv to alert the API when the password changes, but we can implement a way for the API to read their password. We can turn on the GETPASS command, which allows users to retrieve their passwords. The API can then run checkAuthentication against XMLRPC on login, if it doesn't match, it can retrieve the new password from XMLRPC. This, however, requires our nickserv database to not be encrypted.

This is of course somewhat of a security issue, storing plaintext passwords in a DB. But at this point I do believe the API does the same, so it's not a lessening of security.

Another question is of course WHY do we need the API to sync the password in reverse? Do we just let them change their nickserv password individually? Do we enforce the API password being used as their nickserv password?

**Comment by trezy** [comment]
@kenneaal The API passwords are encrypted. We store the salt so we could just hash the password we get back from NickServ and compare but I don't like the idea of leaving the NickServ passwords unencrypted. Is there a way to enforce using the API password as their NS pass? That would be my preference.

**Comment by kenneaal** [comment]
That's exactly what it's doing. But it can't accept an encrypted password over XMLRPC; it has to be the plaintext password (sent over SSL), which then gets stored in nickserv's database, presently encrypted with SHA256. The question @xlexi was asking was whether we could sync the other way around; from the nickserv password being changed on IRC.

**Comment by xlexi** [comment]
Storing passwords in plain text is out of the question, I think I'm in favour of disabling SET password all together and letting people only change their password using the API

**Comment by trezy** [comment]
@kenneaal Gotcha, I missed that. I'm with @xlexi. It'd be better to just disable SET from IRC altogether.

**Comment by kenneaal** [comment]
Well, actually - we talked about it in IRC, and the current plan is:

1) When a rat creates their account, their password is sent from the API to the NickServ database.
2) Users can from that point change their password as they please through the nickserv commands. The API page retains the ability to force-set their nickserv password if they lose it, or the registration lapses.

It is the simplest, and in the end most user friendly method. It lets those who don't know how to use nickserv safe, while not restricting the ones who want to use the features.

**Comment by trezy** [comment]
You're face is an IRC.

**Comment by kenneaal** [comment]
*Your
